### ISSUE TITLE
Picking a different version of NPM in an offline environment requires the Internet

### ISSUE THREAD

**Issue by dmikusa-pivotal** [cloudfoundry/nodejs-buildpack#54]
Steps to reproduce:

1.) Create a project folder and add a file package.json, use these contents.

```
{
  "name": "node-test",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "start": "node index.js"
  },
  "engines": {
    "node": "0.10.42",
    "npm": "2.14.12"
  },
  "author": "Daniel Mikusa",
  "license": "ISC"
}
```

2.) It doesn't really matter, but you can add this small test app in the file index.js.

```
var http = require('http');

http.createServer(function (req, res) {
      res.writeHead(200, {'Content-Type': 'text/plain'});
      for (var env in process.env) {
          res.write(env + '=' + process.env[env] + '\n');
      }
      res.end();
}).listen(process.env.PORT || 3000);
```

3.) Push the application to a CF installation that does not have Internet access. An easy way to simulate this is by unbinding all the staging security groups.

Staging will fail as the build pack runs `npm` to install the requested version of NPM, and NPM needs the Internet to do that.

https://github.com/cloudfoundry/nodejs-buildpack/blob/master/lib/binaries.sh#L61-L62

```
$ cf push node-test-2 -i 1 -m 128M -b node_js_buildpack_v157
Updating app node-test-2 in org lab / space dmikusa as admin...
OK

Uploading node-test-2...
Uploading app files from: /Users/danielmikusa/Downloads/node-test
Uploading 783B, 2 files
Done uploading
OK

Starting app node-test-2 in org lab / space dmikusa as admin...
Downloading node_js_buildpack_v157...
Downloaded node_js_buildpack_v157
Creating container
Successfully created container
Downloading app package...
Downloaded app package (663B)
Staging...
-------> Buildpack version 1.5.7
-----> Creating runtime environment

       NPM_CONFIG_LOGLEVEL=error
       NPM_CONFIG_PRODUCTION=true
       NODE_ENV=production
       NODE_MODULES_CACHE=true
-----> Installing binaries
       engines.node (package.json):  0.10.42
       engines.npm (package.json):   2.14.12

       Downloading and installing node 0.10.42...
       Downloaded [file:///tmp/buildpacks/4885c7f3934d5c1fd6e86a7a917058b5/dependencies/https___pivotal-buildpacks.s3.amazonaws.com_concourse-binaries_node_node-0.10.42-linux-x64.tgz]
       Downloading and installing npm 2.14.12 (replacing version 1.4.29)...
       npm WARN deprecated This version of npm lacks support for important features,
       npm WARN deprecated such as scoped packages, offered by the primary npm
       npm WARN deprecated registry. Consider upgrading to at least npm@2, if not the
       npm WARN deprecated latest stable version. To upgrade to npm@2, run:
       npm WARN deprecated
       npm WARN deprecated   npm -g install npm@latest-2
       npm WARN deprecated
       npm WARN deprecated To upgrade to the latest stable version, run:
       npm WARN deprecated
       npm WARN deprecated   npm -g install npm@latest
       npm WARN deprecated
       npm WARN deprecated (Depending on how Node.js was installed on your system, you
       npm WARN deprecated may need to prefix the preceding commands with `sudo`, or if
       npm WARN deprecated on Windows, run them from an Administrator prompt.)
       npm WARN deprecated
       npm WARN deprecated If you're running the version of npm bundled with
       npm WARN deprecated Node.js 0.10 LTS, be aware that the next version of 0.10 LTS
       npm WARN deprecated will be bundled with a version of npm@2, which has some small
       npm WARN deprecated backwards-incompatible changes made to `npm run-script` and
       npm WARN deprecated semver behavior.
       npm ERR! network getaddrinfo ENOTFOUND
       npm ERR! network This is most likely not a problem with npm itself
       npm ERR! network and is related to network connectivity.
       npm ERR! network In most cases you are behind a proxy or have bad network settings.
       npm ERR! network
       npm ERR! network If you are behind a proxy, please make sure that the
       npm ERR! network 'proxy' config is set properly.  See: 'npm help config'

       npm ERR! System Linux 3.19.0-49-generic
       npm ERR! command "/tmp/app/.heroku/node/bin/node" "/tmp/app/.heroku/node/bin/npm" "install" "--unsafe-perm" "--quiet" "-g" "npm@2.14.12"
       npm ERR! cwd /tmp/app
       npm ERR! node -v v0.10.42
       npm ERR! npm -v 1.4.29
       npm ERR! syscall getaddrinfo
       npm ERR! code ENOTFOUND
       npm ERR! errno ENOTFOUND
       npm ERR!
       npm ERR! Additional logging details can be found in:
       npm ERR!     /tmp/app/npm-debug.log
       npm ERR! not ok code 0
-----> Build failed

       We're sorry this build is failing! You can troubleshoot common issues here:
       https://devcenter.heroku.com/articles/troubleshooting-node-deploys

       If you're stuck, please submit a ticket so we can help:
       https://help.heroku.com/

       Love,
       Heroku

Failed to compile droplet
Exit status 223
Staging failed: Exited with status 223

FAILED
BuildpackCompileFailed

TIP: use 'cf logs node-test-2 --recent' for more information
```

I don't believe that vendoring will help here as the build pack is trying to install NPM with the `-g` flag, which installs it globally and not into the vendored dependencies.

**Comment by Dannyzen** [comment]
@dmikusa-pivotal thanks for putting in this issue. We're curious as to what you perceive expected behavior to be in this use case.

**Comment by dmikusa-pivotal** [comment]
Well given that it's an offline / cached build pack, I would not expect it to try to access the Internet, which it's doing.  I say that the build pack is doing this not NPM because it's the build pack that's executing an NPM command that has no hope of succeeding without Internet access.  

I'm not sure exactly what the right behavior is here, but it shouldn't be to try accessing the Internet as that's a guaranteed failure in an offline environment.

I can see a couple different possibilities for handling this nicer when the build pack is in offline mode.  For either case it basically involves not allowing the user to select a custom version of NPM (a case we can detect).

The first option, if the user tried to select a custom version of NPM, we would log an error and print a clear message to the user saying that we cannot accommodate the request for a custom NPM version as it would require access to the internet to install that version of NPM.  Linking to a doc that explains more would be helpful, then we could fully explain the customer's options (like how to remove custom NPM version or how to use an online build pack with an HTTP_PROXY).

The second option would be to simply skip installing the requested version and produce a big warning message that says you're not getting the version of NPM you selected because we're in offline mode and it requires the Internet to download and install that version of NPM.  Again, a link to the docs with more info on this would be helpful too.

I'm open to other suggestions as well.  Those were just the first two that came to mind.

**Comment by flavorjones** [comment]
I agree. If it needs to hit the internet, I'd expect us to emit a not-in-the-manifest error similar to what happens if I specify an unavailable version of node.

**Comment by jvshahid** [comment]
Just few clarifications. the buildpack doesn't have `npm` in the manifest. It simply uses the `npm` that is packaged with the version of `node` you specify in the `package.json` file. I believe heroku added this feature to their buildpack since npm prior to `2.1.x` had major bugs in them. The only version of node that ships with such an old version of `npm` are the `0.10.x`. Related to the previous note, I think it is better to print a warning if the `npm` version doesn't match the one bundled with nodejs. Erroring out has the potential of breaking "many" apps which relied on this behavior.

**Comment by jvshahid** [comment]
A pointer to heroku's documentation: https://devcenter.heroku.com/articles/nodejs-support#specifying-an-npm-version

**Comment by jvshahid** [comment]
Sent a cf-dev proposal to change the buildpack behavior. the new behavior will be to not allow developers to change the version of npm. as a developer you can only use the version of npm that is shipped with the version of nodejs you chose. You can find the discussion here: https://lists.cloudfoundry.org/archives/list/cf-dev@lists.cloudfoundry.org/thread/ZRSOFXISIR4PHWIRFN4RABBRAGE42VZR/

**Comment by jvshahid** [comment]
@flavorjones do you still have strong opinions about this behavior given the feedback that we got in the mailing list and here ?

**Comment by flavorjones** [comment]
@jvshahid I never had a strong opinion about this, but it's not clear what the proposal is we're currently considering. Can you summarize here?

**Comment by jvshahid** [comment]
Modify the buildpack to *not* download a different version of `npm` if it isn't specified in `package.json`. Also, modify the documentation to clearly states that you shouldn't specify `npm` version in `package.json` if you are in an offline environment.

**Comment by tomoe** [comment]
@jvshahid 

> modify the documentation to clearly states that you shouldn't specify npm version in package.json if you are in an offline environment.

I would also like to see an error message logged and printed out to the user as @dmikusa-pivotal suggested in https://github.com/cloudfoundry/nodejs-buildpack/issues/54#issuecomment-202434260

**Comment by Dannyzen** [comment]
There's quite a few different issues & stories related to this originating issue. The below describes the actions the team will take:

* Document that custom versions of npm, in an offline environment are not supported
* Add a test to make sure that if no version of npm is selected, we use the default version that comes with node and no internet traffic attempts are made

**Comment by Dannyzen** [comment]
This work has been completed on: https://www.pivotaltracker.com/story/show/118775597
